File: android/app/src/main/kotlin/app/zen/chromium/ext/SystemInfo.kt
package app.zen.chromium.ext

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * `chrome.system.cpu` and `chrome.system.memory` on the phone (`ext.system.cpu`,
 * `ext.system.memory`): the readings the core shapes into Chrome's `CpuInfo` and `MemoryInfo`
 * (`core/extensions/api/systemInfo.ts`, which also gates them on the permissions).
 *
 * The processors: `Runtime.availableProcessors()` for the count, Java's `os.arch` for the
 * architecture (the kernel's machine name, what Chrome's `SysInfo` reads), `/proc/cpuinfo` for
 * the model (`model name` on x86, the `Hardware` line on ARM as `base::CPU` falls back to) and
 * the x86 feature flags, `/proc/stat` for the per-processor times. The app's sandbox keeps
 * `/proc/stat` from an untrusted app since Android 8 and may keep `/proc/cpuinfo` too; what
 * cannot be read is left out (`usage` null, the model from `Build`), and the core fills the
 * shape with zeros. The memory: `ActivityManager.MemoryInfo.totalMem` and `availMem`, bytes.
 */
object SystemInfo {
    /** The x86 features Chrome's `CpuInfoProvider` lists, in its order; Linux spells SSE3 `pni`. */
    private val FEATURES = listOf("mmx", "sse", "sse2", "sse3", "ssse3", "sse4_1", "sse4_2", "avx")

    /** The model and the features a `/proc/cpuinfo` text names. */
    class CpuText(val modelName: String, val features: List<String>)

    /** `/proc/cpuinfo`: the first `model name` (x86) or `Hardware` (ARM) as the model, the first `flags` line as the features. */
    fun parseCpuInfo(text: String): CpuText {
        var modelName = ""
        var hardware = ""
        var features: List<String> = emptyList()
        for (line in text.lineSequence()) {
            val colon = line.indexOf(':')
            if (colon < 0) continue
            val key = line.substring(0, colon).trim().lowercase()
            val value = line.substring(colon + 1).trim()
            when {
                key == "model name" && modelName.isEmpty() -> modelName = value
                key == "hardware" && hardware.isEmpty() -> hardware = value
                key == "flags" && features.isEmpty() -> features = featuresOf(value)
            }
        }
        return CpuText(modelName.ifEmpty { hardware }, features)
    }

    /** Chrome's feature names present in a `flags` line, in Chrome's order. */
    fun featuresOf(flags: String): List<String> {
        val present = flags.split(Regex("\\s+")).map { it.lowercase() }.toHashSet()
        if ("pni" in present) present.add("sse3")
        return FEATURES.filter { it in present }
    }

    /**
     * `/proc/stat`: each `cpuN` line's `[user + nice, system, idle]` in processor order as Chrome's
     * Linux provider reads them; a processor missing from the numbering reads as no time counted.
     * Null for a text with no `cpuN` line (the sandbox's `Permission denied`, an empty read).
     */
    fun parseProcStat(text: String): List<LongArray>? {
        val byIndex = HashMap<Int, LongArray>()
        var max = -1
        for (line in text.lineSequence()) {
            val m = STAT_LINE.matchEntire(line.trimEnd()) ?: continue
            val index = m.groupValues[1].toInt()
            val user = m.groupValues[2].toLong() + m.groupValues[3].toLong()
            byIndex[index] = longArrayOf(user, m.groupValues[4].toLong(), m.groupValues[5].toLong())
            if (index > max) max = index
        }
        if (max < 0) return null
        return List(max + 1) { byIndex[it] ?: longArrayOf(0, 0, 0) }
    }

    private val STAT_LINE = Regex("^cpu(\\d+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)(?:\\s.*)?$")

    /** The reading as the core takes it: `{ numOfProcessors, archName, modelName, features, usage }`. */
    fun cpuReading(
        processors: Int,
        arch: String,
        cpuInfoText: String?,
        procStatText: String?,
        fallbackModel: String
    ): JSONObject {
        val parsed = cpuInfoText?.let { parseCpuInfo(it) }
        val usage = procStatText?.let { parseProcStat(it) }
        val out = JSONObject()
            .put("numOfProcessors", processors.coerceAtLeast(1))
            .put("archName", arch)
            .put("modelName", parsed?.modelName?.ifEmpty { fallbackModel } ?: fallbackModel)
            .put("features", JSONArray(parsed?.features ?: emptyList<String>()))
        if (usage == null) out.put("usage", JSONObject.NULL)
        else out.put("usage", JSONArray().also { array -> usage.forEach { array.put(JSONArray().put(it[0]).put(it[1]).put(it[2])) } })
        return out
    }

    /** The phone's processors, read now (file IO: off the main thread). */
    fun cpu(): JSONObject = cpuReading(
        Runtime.getRuntime().availableProcessors(),
        System.getProperty("os.arch") ?: "",
        readProc("/proc/cpuinfo"),
        readProc("/proc/stat"),
        socModel()
    )

    /** The phone's memory as `ActivityManager` reports it: `{ capacity, availableCapacity }`, bytes. */
    fun memory(context: Context): JSONObject {
        val info = ActivityManager.MemoryInfo()
        (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager).getMemoryInfo(info)
        return JSONObject().put("capacity", info.totalMem).put("availableCapacity", info.availMem)
    }

    /** The SoC's model when the build names it (API 31), else the board's hardware name. */
    private fun socModel(): String {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val soc = Build.SOC_MODEL
            if (soc.isNotEmpty() && soc != Build.UNKNOWN) return soc
        }
        return Build.HARDWARE.takeIf { it != Build.UNKNOWN } ?: ""
    }

    /** A `/proc` file's text, or null where the sandbox refuses it. */
    private fun readProc(path: String): String? = runCatching { File(path).readText() }.getOrNull()
}

File: src/core/extensions/api/systemInfo.ts
/**
 * `chrome.system.cpu` and `chrome.system.memory` in Chrome's shape (`CpuInfoProvider`,
 * `MemoryInfoProvider`), over what each host reads: the phone's `Runtime` / `ActivityManager`
 * and its `/proc` files where the app may read them, the desktop's `os` module. Chrome hides
 * each namespace from extensions that do not declare its permission; so does the shim, and the
 * hosts check again. Pure shaping here, so every host answers the same fields.
 */

export const SYSTEM_CPU_PERMISSION = 'system.cpu'
export const SYSTEM_MEMORY_PERMISSION = 'system.memory'

export const SYSTEM_CPU_NO_PERMISSION_ERROR =
  "The extension does not have the 'system.cpu' permission."
export const SYSTEM_MEMORY_NO_PERMISSION_ERROR =
  "The extension does not have the 'system.memory' permission."

/** Chrome's `system.cpu.CpuTime`: cumulative ticks, `total` the sum of the three. */
export interface CpuTime {
  user: number
  kernel: number
  idle: number
  total: number
}

/** Chrome's `system.cpu.CpuInfo`. */
export interface CpuInfo {
  numOfProcessors: number
  archName: string
  modelName: string
  features: string[]
  processors: Array<{ usage: CpuTime }>
  /** ChromeOS reads its thermal zones; every other platform answers none. */
  temperatures: number[]
}

/** Chrome's `system.memory.MemoryInfo`, both in bytes. */
export interface MemoryInfo {
  capacity: number
  availableCapacity: number
}

/** What a host reads for `system.cpu.getInfo`; the shaping below fills in what it left out. */
export interface RawCpuReading {
  numOfProcessors?: unknown
  archName?: unknown
  modelName?: unknown
  features?: unknown
  /** Per processor `[user, kernel, idle]` ticks, `user` with `nice` folded in as Chrome folds it. */
  usage?: unknown
}

/** What a host reads for `system.memory.getInfo`. */
export interface RawMemoryReading {
  capacity?: unknown
  availableCapacity?: unknown
}

/** The x86 features Chrome's `CpuInfoProvider` lists, in its order (ARM lists none). */
export const CPU_FEATURES = [
  'mmx',
  'sse',
  'sse2',
  'sse3',
  'ssse3',
  'sse4_1',
  'sse4_2',
  'avx'
] as const

/**
 * Chrome's `archName`: `base::SysInfo::OperatingSystemArchitecture()`, the machine name with
 * the i?86 family spelled `x86` and `amd64` spelled `x86_64`; `aarch64`, `armv7l`, `armv8l` as
 * the kernel (or Java's `os.arch`) spells them. Node's `x64` / `ia32` / `arm` are mapped too.
 */
export function chromeArchName(raw: string): string {
  const arch = raw.trim()
  if (/^i[3-6]86$/.test(arch) || arch === 'ia32' || arch === 'x86') return 'x86'
  if (arch === 'amd64' || arch === 'x64') return 'x86_64'
  return arch
}

/**
 * The features Chrome lists from a `/proc/cpuinfo` `flags` line (Linux spells SSE3 `pni`); an
 * ARM `Features` line yields none, as Chrome's ARM build lists none.
 */
export function cpuFeaturesFromFlags(flags: string): string[] {
  const present = new Set(flags.split(/\s+/).map((flag) => flag.toLowerCase()))
  if (present.has('pni')) present.add('sse3')
  return CPU_FEATURES.filter((feature) => present.has(feature))
}

const count = (value: unknown): number =>
  typeof value === 'number' && Number.isFinite(value) && value > 0 ? Math.floor(value) : 0

const ticks = (value: unknown): number =>
  typeof value === 'number' && Number.isFinite(value) && value >= 0 ? Math.floor(value) : 0

/** One processor's `CpuTime` from `[user, kernel, idle]`; anything else reads as no time counted. */
export function cpuTime(entry: unknown): CpuTime {
  const parts = Array.isArray(entry) ? entry : []
  const user = ticks(parts[0])
  const kernel = ticks(parts[1])
  const idle = ticks(parts[2])
  return { user, kernel, idle, total: user + kernel + idle }
}

/**
 * A host's reading shaped as Chrome's `CpuInfo`: as many `processors` entries as
 * `numOfProcessors` (a host that could not read the per-processor times, the phone's app
 * sandbox keeps `/proc/stat` from it, gives them all as zero), the architecture in Chrome's
 * spelling, the model name a string, the features Chrome's list in Chrome's order.
 */
export function cpuInfo(raw: RawCpuReading): CpuInfo {
  const usage = Array.isArray(raw.usage) ? raw.usage : []
  const numOfProcessors = Math.max(count(raw.numOfProcessors), usage.length, 1)
  const processors: Array<{ usage: CpuTime }> = []
  for (let i = 0; i < numOfProcessors; i++) processors.push({ usage: cpuTime(usage[i]) })
  const listed: unknown[] = Array.isArray(raw.features) ? raw.features : []
  const features = CPU_FEATURES.filter((feature) => listed.includes(feature))
  return {
    numOfProcessors,
    archName: chromeArchName(typeof raw.archName === 'string' ? raw.archName : ''),
    modelName: typeof raw.modelName === 'string' ? raw.modelName.trim() : '',
    features: [...features],
    processors,
    temperatures: []
  }
}

/** A host's reading shaped as Chrome's `MemoryInfo`: bytes, the available never over the capacity. */
export function memoryInfo(raw: RawMemoryReading): MemoryInfo {
  const capacity = ticks(raw.capacity)
  const availableCapacity = Math.min(ticks(raw.availableCapacity), capacity)
  return { capacity, availableCapacity }
}

/**
 * The per-processor times of a Linux `/proc/stat`: each `cpuN` line's `user nice system idle`
 * as Chrome's `CpuInfoProvider` reads it (`user + nice`, `system`, `idle`), in processor order;
 * an unreadable or foreign text yields none.
 */
export function procStatUsage(text: string): Array<[number, number, number]> {
  const out: Array<[number, number, number]> = []
  for (const line of text.split('\n')) {
    const match = /^cpu(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)/.exec(line)
    if (!match) continue
    out[Number(match[1])] = [
      Number(match[2]) + Number(match[3]),
      Number(match[4]),
      Number(match[5])
    ]
  }
  for (let i = 0; i < out.length; i++) out[i] ??= [0, 0, 0]
  return out
}

/**
 * What a Linux `/proc/cpuinfo` says of the model and the features: the first `model name`
 * (x86) or `Hardware` (ARM, as `base::CPU` falls back to) as the model, the first `flags` line
 * as the features. Either may be missing; the phone's sandbox may keep the file itself away.
 */
export function procCpuInfo(text: string): { modelName: string; features: string[] } {
  let modelName = ''
  let hardware = ''
  let features: string[] = []
  for (const line of text.split('\n')) {
    const colon = line.indexOf(':')
    if (colon < 0) continue
    const key = line.slice(0, colon).trim().toLowerCase()
    const value = line.slice(colon + 1).trim()
    if (key === 'model name' && !modelName) modelName = value
    else if (key === 'hardware' && !hardware) hardware = value
    else if (key === 'flags' && features.length === 0) features = cpuFeaturesFromFlags(value)
  }
  return { modelName: modelName || hardware, features }
}
